js_modules/dagit/packages/core/src/ui/Checkbox.tsx

```diff
@@ -0,0 +1,221 @@
+      rx="10"
+      fill={ColorsWIP.White}
+    />
+  </svg>
+);
+
+const CheckIcon: React.FC<IconProps> = ({checked, indeterminate, fillColor}) => (
+  <svg width="24px" height="24px" viewBox="-3 -3 24 24">
+    <path
+      d="M16,0 C17.1,0 18,0.9 18,2 L18,2 L18,16 C18,17.1 17.1,18 16,18 L16,18 L2,18 C0.9,18 0,17.1 0,16 L0,16 L0,2 C0,0.9 0.9,0 2,0 L2,0 Z"
+      id="Background"
+      className=" interaction-focus-outline"
+      style={{transition: 'fill 100ms linear'}}
+      fill={ColorsWIP.White}
+    />
+    <path
+      id="Border"
+      className="interaction-darken"
+      d="M15 16H3C2.45 16 2 15.55 2 15V3C2 2.45 2.45 2 3 2H15C15.55 2 16 2.45 16 3V15C16 15.55 15.55 16 15 16ZM16 0H2C0.9 0 0 0.9 0 2V16C0 17.1 0.9 18 2 18H16C17.1 18 18 17.1 18 16V2C18 0.9 17.1 0 16 0Z"
+      fill={ColorsWIP.Gray300}
+    />
+    <path
+      d="M16,0 C17.1,0 18,0.9 18,2 L18,2 L18,16 C18,17.1 17.1,18 16,18 L16,18 L2,18 C0.9,18 0,17.1 0,16 L0,16 L0,2 C0,0.9 0.9,0 2,0 L2,0 Z"
+      id="Fill"
+      className="interaction-darken"
+      style={{transition: 'fill 100ms linear'}}
+      fill={checked || indeterminate ? fillColor : 'transparent'}
+    />
+    <polyline
+      id="Check"
+      fill="none"
+      stroke={ColorsWIP.White}
+      strokeWidth="2"
+      strokeLinecap="round"
+      strokeLinejoin="round"
+      strokeDasharray="16"
+      style={{
+        transition: 'stroke-dashoffset 100ms linear',
+        transitionDelay: !(checked && !indeterminate) ? '80ms' : '0',
+      }}
+      strokeDashoffset={checked && !indeterminate ? '0' : '16'}
+      points="3.5 9 7 12.5 14.5 5.0"
+    />
+    <line
+      id="Indeterminate"
+      x1="5"
+      y1="9"
+      x2="13"
+      y2="9"
+      style={{
+        transition: 'stroke-dashoffset 100ms linear',
+        transitionDelay: !indeterminate ? '80ms' : '0',
+      }}
+      stroke={ColorsWIP.White}
+      strokeWidth="2"
+      strokeDasharray="8"
+      strokeLinecap="round"
+      strokeDashoffset={indeterminate ? '0' : '8'}
+    />
+  </svg>
+);
+
+export const Checkbox = styled(
+  ({
+    id,
+    checked,
+    label,
+    className,
+    format = 'check',
+    disabled = false,
+    indeterminate = false,
+    fillColor = ColorsWIP.Gray800,
+    ...rest
+  }) => {
+    const uid = useRef(id || uniqueId('checkbox-'));
+    const Component: React.FC<IconProps> = {star: StarIcon, check: CheckIcon, switch: SwitchIcon}[
+      format
+    ];
+
+    return (
+      <label htmlFor={uid.current} className={className}>
+        <input
+          type="checkbox"
+          id={uid.current}
+          tabIndex={0}
+          checked={checked}
+          disabled={disabled}
+          indeterminate={indeterminate}
+          {...rest}
+        />
+        <Component
+          disabled={disabled}
+          checked={checked}
+          indeterminate={indeterminate}
+          fillColor={disabled ? DISABLED_COLOR : fillColor}
+        />
+        {label}
+      </label>
+    );
+  },
```

Comment:
Any reason not to define the wrapped component separately? I find this a bit harder to follow. Also, without knowing how styled-components works internally, is it possible that this component definition is recreated every time a `Checkbox` is used?

(I've seen similar cases where HOCs that use inline component definitions end up doing wayyy too much and leading to bugs that are super tough to track down.)

Reply:
Ok I'll break them out! I don't have a preference, just thought it was kind of nice to avoid having to name it separately. 

I'm not sure about the definition being recreated - I wouldn't think so since this isn't adding an extra layer of `() =>` over referencing a function defined as a variable